### PR TITLE
Feature/typescript

### LINKED ISSUE #14: Investigate feasibility/benefit of porting project to Typescript
**Description**
Investigate how much of a lift porting the project over to Typescript would be. Document any caveats or potential roadblocks, changes in development workflow, npm package publishing, etc.

### PR DESCRIPTION
**Pull request by brandongregoryscott**
Closes #14 Investigate feasibility/benefit of porting project to Typescript 

Initial pass of the TS port for the project. All automated tests should be updated & passing. There are likely some missed areas of type narrowing/specification that can be filled in later.

A few things of note:
- I opted to *not* use `ts-node` for development. I was unable to use _just_ `ts-node` to run anything but the main entrypoint file, and subcommands were not registered. It required a much longer command string pointing to the tsconfig file (see https://github.com/tj/commander.js/issues/955#issuecomment-490016192 for an example)
    - With that in mind, while developing, it'll be the smoothest experience to run the `watch` npm script as the `and-cli-dev` alias now points to `<repo>/dist/and-cli.js`
    - Integration test files were converted over to `.ts`, but the `TestUtils` function that executes the CLI commands will also reference the `dist` folder for executable js files.
- Jest was slightly upgraded (to the latest released version of v25), but `ts-jest` is throwing warnings about using jest <26. I believe jest v26+ requires node 10 and above, so we'd need to agree to move forward with that upgrade as a team - there are some implications with our Jenkins server that is still on node 8.16.3
- All modules were capitalized to follow our standard convention with utility classes (`CoreUtils`, `CollectionUtils`, etc.). So any plugin projects or older branches referencing them will need to be updated.
- Additionally, the main entrypoint (`and-cli.ts`) will now export all of the shared modules for consumer use, similar to how our `andculturecode-javascript-core` and `andculturecode-javascript-react` packages export everything in the `index.ts` file.

@wintondeshong This is a massive PR, so I don't expect you to scan everything line-by-line (some of the restructuring caused git to register the changes as new files entirely), but I'd love to get your input on some of this, too. In terms of releasing it, I was thinking we could release it as `2.0.0-beta` version(s) until we iron out any of the potential regressions with the port - what do you think? 

-   [x] Related GitHub issue(s) linked in PR description
-   [x] Destination branch merged, built and tested with your changes
-   [x] Code formatted and follows best practices and patterns
-   [x] Code builds cleanly (no _additional_ warnings or errors)
    - See above notes about the `ts-jest` warnings
-   [x] Manually tested
-   [x] Automated tests are passing
-   [-] No _decreases_ in automated test coverage
    - Codecov is reading a small decrease in %, but there shouldn't be a ton of new code. We can revisit the coverage after the initial port is in.
-   [x] Documentation updated (readme, docs, comments, etc.)
-   [x] Localization: No hard-coded error messages in code files (_minimally_ in string constants)

<details>
<summary>Instructions for beta-testing</summary>

```
# Ensure latest and-cli is installed for access to workspace command
npm i -g and-cli@latest

# Ensure you are in a directory appropriate for holding OSS repos (can be anywhere, just an example)
cd ~/andculturecode/forks/

# Clone all of my forks
and-cli workspace -u brandongregoryscott && cd brandongregoryscott.AndcultureCode.Cli

# Fetch remote branches and checkout the typescript branch
git fetch && git checkout feature/typescript

# Install dependencies and run a build
npm install && npm run build

# Install the CLI aliases
./dist/and-cli.js install

Now you can run the built JS version of the CLI by running `and-cli-dev`
```
</details>

<details>
<summary>Command audit/testing checklist</summary>

- [x] `copy`
    - [x] `-d, --destination <destination>`
    - [x] `-f, --flags <options>`
    - [x] `-s, --source <source>`
- [x] `deploy`
    - [x] `aws-beanstalk`
        - [x] `--dotnet`
        - [x] `--timeout <timeout>`
        - [x] `--verbose`
    - [x] `aws-s3`
        - [x] `--destination <destination>`
        - [x] `--profile <profile>`
        - [x] `--public-url <url>`
        - [x] `--publish`
        - [x] `--source <source>`
        - [x] `--webpack`
    - [x] `azure-storage`
        - [x] `--client-id <clientID>`
        - [x] `--destination <destination>`
        - [x] `--public-url <url>`
        - [x] `--publish`
        - [x] `--recursive`
        - [x] `--secret <profile>`
        - [x] `--source <source>`
        - [x] `--tenant-id <tenantID>`
        - [x] `--username <username>`
        - [x] `--webpack`
    - [x] `azure-web-app`
        - [x] `--app-name <applicationName>`
        - [x] `--branch <branch>`
        - [x] `--client-id <clientID>`
        - [x] `--force`
        - [x] `--remote <remote>`
        - [x] `--resource-group <resourceGroup>`
        - [x] `--secret <profile>`
        - [x] `--tenant-id <tenantID>`
        - [x] `--username <username>`
    - [x] `jenkins`
        - [x] `--new`
        - [x] `-i, --init`
        - [x] `-p, --profile <profile>`
- [x] `dotnet`
    - [x] `-b, --build`
    - [x] `-c, --clean`
    - [x] `-C, --cli`
    - [x] `-k, --kill`
    - [x] `-p, --publish`
    - [x] `-R, --restore`
    - [x] `-r, --run`
    - [x] `-w, --watch`
- [x] `dotnet-test`
    - [x] `--by-project`
    - [x] `--ci`
    - [x] `--coverage`
    - [x] `-s, --skip-clean`
- [x] `github`
    - [x] `--add-topic <topic>`
    - [x] `--list-repos`
    - [x] `--list-topics`
    - [x] `--remove-topic <topic>`
    - [x] `-r, --repo <repo>`
    - [x] `-u, --username <username>`
- [x] `install`
- [x] `migration`
    - [x] `-a, --add`
    - [x] `-d, --delete`
    - [x] `-r, --run`
- [x] `nuget`
    - [x] `-p, --publish <version>`
    - [x] `-u, --upgrade`
- [x] `webpack`
    - [x] `-c, --clean` 
    - [x] `-p, --publish`
    - [x] `-R, --restore`
- [x] `webpack-test`
    - [x] `--ci`
    - [x] `-c, --clean`
    - [x] `-R, --restore`
- [x]  `workspace`
    - [x] `-c, --clone`
    - [x] `-f, --fork`
    - [x] `-u, --usernames <usernames>`

</details>

### CODE REVIEW

**Comment by wintondeshong** [comment]
@brandongregoryscott going to try this out this afternoon, but after reading it through it looks awesome! phew, nice work 👏

File: src/and-cli-install.ts
Comment (wintondeshong) [inline src/and-cli-install.ts:101]: 👍 good attention to detail on the install command needing to point to dist. -- in reading this again, I'm thinking we could update this function (in a later PR) to support passing the suffix (ie. dev, etc...). as a maintainer, I keep aliases that point to each of the contributors of this project (ie. `and-cli-{github-username}`). Wondering if the workspace command could also setup/sync aliases too.

File: src/and-cli-workspace.ts
Comment (wintondeshong) [inline src/and-cli-workspace.ts:1]: @brandongregoryscott I can't quite spot the breaking change without digging in further, but I am noticing a regression with the `workspace` command's exiting. No matter the variation (even `--help`) it hangs and I can't even `ctrl+c` to exit.

Thing is... it eventually does exit. Just takes _several minutes_.

File: src/modules/github.ts
Comment (wintondeshong) [inline src/modules/github.ts:1]: So upon further testing of the issue with exits hanging, it appears to be effecting both the `github` and `workspace` command series. The common denominator of course is the `github` module. There is likely something in here causing the exits to hang. It does eventually exit. Just takes several minutes.

**Review by wintondeshong — CHANGES_REQUESTED** [review]
@brandongregoryscott knocked out testing a slew of the commands above. I've updated the checklist. I did catch a regression with what I believe is likely happening in the github module. It is likely something simple, but feel it should be investigated before merge.

Totally cool with a beta version 👍 perhaps this branch would end up being merged into a beta branch that we publish off of until we are ready to merge into main.

File: src/and-cli-install.ts
Comment (brandongregoryscott) [reply src/and-cli-install.ts:101]: That would be awesome! I'll add an issue to document this

File: src/modules/github.ts
Comment (brandongregoryscott) [reply src/modules/github.ts:1]: Thanks for digging into that 👍 I'll take a further look

File: src/and-cli-workspace.ts
Comment (brandongregoryscott) [reply src/and-cli-workspace.ts:1]: Weird. I've been testing this out this morning and I did not see the slowness with the `github` command, but I did with the `workspace` command. I even stripped out almost everything (functions, imports, etc) from that command and it still seems to hang for ~30s to display the help menu... 🤔

**Comment by brandongregoryscott** [comment]
@wintondeshong I believe (at least part of) the slowdown is from the exports in `and-cli.ts`. I added those in pretty late in the port which is probably why I didn't notice it. If you comment out all of those, rebuild and run `and-cli-dev github`, the time goes from ~5s to ~1s:

```
bscot@BSCOTT-PC ~/acc/fork/brandongregoryscott.AndcultureCode.Cli/ (feature/typescript) -> time and-cli-dev github
Usage: and-cli-github option

Helpful Github operations used at andculture

Options:
  --add-topic <topic>        Add topic to specified repository, or all AndcultureCode repositories if no repo provided
  --list-repos               Lists all AndcultureCode repos
  --list-topics              List topics for a given repo (used in conjunction with --repo)
  --remove-topic <topic>     Remove topic from specified repository, or all AndcultureCode repositories if no repo provided
  -r, --repo <repo>          Repository name to act on (used in conjunction with --add-topic or --list-topics, for example).
  -u, --username <username>  Github username for which to list AndcultureCode repositories
  -h, --help                 display help for command

real    0m4.320s
user    0m0.015s
sys     0m0.015s
```

```
bscot@BSCOTT-PC ~/acc/fork/brandongregoryscott.AndcultureCode.Cli/ (feature/typescript) -> time and-cli-dev github
Usage: and-cli-github option

Helpful Github operations used at andculture

Options:
  --add-topic <topic>        Add topic to specified repository, or all AndcultureCode repositories if no repo provided
  --list-repos               Lists all AndcultureCode repos
  --list-topics              List topics for a given repo (used in conjunction with --repo)
  --remove-topic <topic>     Remove topic from specified repository, or all AndcultureCode repositories if no repo provided
  -r, --repo <repo>          Repository name to act on (used in conjunction with --add-topic or --list-topics, for example).
  -u, --username <username>  Github username for which to list AndcultureCode repositories
  -h, --help                 display help for command

real    0m1.130s
user    0m0.015s
sys     0m0.015s
```

...which sucks. I was hoping to improve the consumer experience by exporting all of the available modules, but maybe there's another approach we can take.

**Comment by wintondeshong** [comment]
> @wintondeshong I believe (at least part of) the slowdown is from the exports in `and-cli.ts`. I added those in pretty late in the port which is probably why I didn't notice it. If you comment out all of those, rebuild and run `and-cli-dev github`, the time goes from ~5s to ~1s:
> 
> ```
> bscot@BSCOTT-PC ~/acc/fork/brandongregoryscott.AndcultureCode.Cli/ (feature/typescript) -> time and-cli-dev github
> ...
> 
> real    0m4.320s
> user    0m0.015s
> sys     0m0.015s
> ```
> 
> ```
> bscot@BSCOTT-PC ~/acc/fork/brandongregoryscott.AndcultureCode.Cli/ (feature/typescript) -> time and-cli-dev github
> ...
> 
> real    0m1.130s
> user    0m0.015s
> sys     0m0.015s
> ```
> 
> ...which sucks. I was hoping to improve the consumer experience by exporting all of the available modules, but maybe there's another approach we can take.
> 
> Update: moved all of the exports to `index.ts` which is the new "main" file for the package. This should allow us to have easy imports + comparable speed to the pre-TS commands

I'm going to step through this here because mine isn't a matter of seconds. It finishes running the actual work from the command (ie. `workspace -u wintondeshong`) and then hangs for upwards of 30 seconds to a minute.

**Comment by brandongregoryscott** [comment]
> I'm going to step through this here because mine isn't a matter of seconds. It finishes running the actual work from the command (ie. `workspace -u wintondeshong`) and then hangs for upwards of 30 seconds to a minute.

Let me know what you can find. If you have some free time this week, we can try to pair up on it. I have not seen _that_ level of slowness on my Windows machine or Macbook, before or after the export changes.

**Comment by wintondeshong** [comment]
Here is my current timing...

**Comment by wintondeshong** [comment]
@brandongregoryscott something seemed up so I re-cloned and did a fresh build of your branch and I'm no longer seeing that performance issue 🤷‍♂️ 

```
05:35 /c/Clients/andculturecode  $ time and-cli-brandongregoryscott workspace -u wintondeshong

[and-cli] -----------------------------------------------------------------
[and-cli]
[and-cli] Configuring workspace
[and-cli]
[and-cli] -----------------------------------------------------------------

[and-cli] Synchronizing forks of AndcultureCode for 'wintondeshong'...

[and-cli] Results
[and-cli]  - Successful: 0
[and-cli]  - Unmodified: 24
[and-cli]  - Errored: 0
[and-cli]  - Total: 24



real    0m2.080s
user    0m0.031s
sys     0m0.000s
```

**Review by wintondeshong — APPROVED** [review]
@brandongregoryscott with the performance issue no longer an issue I'm good to approve. I tested the `install` and `nuget --publish` commands as well, which leaves basically the `deploy` series of commands. Recommend getting someone to test those out who might have them in action presently, which would save us time testing.

**Comment by brandongregoryscott** [comment]
> @brandongregoryscott something seemed up so I re-cloned and did a fresh build of your branch and I'm no longer seeing that performance issue 🤷‍♂️

So weird 🤔 well, glad it's not an issue!

> @brandongregoryscott with the performance issue no longer an issue I'm good to approve. I tested the install and nuget --publish commands as well, which leaves basically the deploy series of commands. Recommend getting someone to test those out who might have them in action presently, which would save us time testing.

Awesome! Thanks for helping out with the testing process. I'll reach out to some of the engineers that are leveraging the remaining deployment commands to see if they have capacity to smoke test the changes.

**Pull request merged**
